docs/fml-spec.md

```diff
@@ -0,0 +1,576 @@
+
+Occassionally, we would like to vary a configuration based upon the device. In this example, the `homepage` feature has sections for both Pocket and Topsites.
+
+```yaml
+features:
+  homepage:
+    variables:
+      pocket-enabled:
+        description: If true, show stories from Pocket. This is only available in certain territories.
+        type: Bool
+        default: false
+      topsites-enabled:
+        description: If true, show tiles from the users most recent and frequently visited pages.
+        type: Bool
+        default: false
+    defaults:
+      - channel: nightly
+        value: { topsites-enabled: true }
+      - channel: nightly
+        targeting: locale in ['en-US', 'de-DE', 'en-GB']
+        value: { pocket-enabled: true }
+```
+
+In this example, `pocket-enabled` defaults to `true` only on the `nightly` channel and when the device locale is in `US`, `DE` or `GB`.
+
+## Additional types
+
+### Primitive types
+
+Primitive types supported:
+
+ * `String`
+ * `Boolean` or `Bool`. Both Kotlin and Swift's spelling is supported.
+ * `Int`
+
+> ⚠️ Unimplemented
+
+ * `Double`
+
+### Types coerced from `String`
+
+> ⚠️ Unimplemented.
+
+We have already seen enums being declared in the `types` section of the manifest.
+
+Other types that may be derived from `String`:
+
+ * `Color`
+ * `URL`
+
+### Suggested values
+
+> ⚠️ Unimplemented.
+
+Strings that can be coerced to resources within the bundle; however they are error prone to type in the Experimenter interface.
+
+Suggested values can be used to suggest values to the experiment owner, with a description on what the resource will look like.
+
+ * `Text` performs a lookup for displayable text in the application bundle. If not text exists, returns the string used for lookup.
+ * `Image` performs a lookup for an image in the application bundle.
+
+For each of these types, the default value *must* correspond to a valid resource identifier in the app.
+
+```yaml
+features:
+    upgrade-message:
+        description: A message displayed when the user upgrades
+        variables:
+            hero-image:
+                description: A pre-bundled image
+                type: Image
+                default: ic_fox
+                suggested:
+                  ic_fox: A picture of a cute fox
+                  ic_snow_fox: A picture of a cute snow fox
+                  ic_redpanda: A picture of a cute red panda
+            message-content:
+                description: Text content to show the user
+                default: msg_thankyou
+                suggested:
+                  msg_welldone: A nice well done for upgrading
+                  msg_thankyou: A nice thank you for sticking with us
+                  msg_warning: A warning that bad things happen to good people
+```
+
+### Object types
+
+Some features require more organization. In the case, JSON Objects can be coerced into generated data classes.
```

Comment:
Is the idea here that you will assume that you have defined `DialogButton somewhere in the application?
How do you distinguish misspelled built-in supported types and custom types?

Also, how does experimenter pick up what to do here?

Reply:
* Experimenter doesn't know how to support these sorts of types. Yet.
* How do you distinguish misspelled built-in supported types and custom types? — it's a good question. For Kotlin, there's package names, but more likely is that you wouldn't call them that. Changed the names in the PR.